pico/tools/src/main/java/io/helidon/pico/tools/ModuleInfoDescriptor.java

```diff
@@ -0,0 +1,633 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.pico.tools;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.StringReader;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+import io.helidon.builder.Builder;
+import io.helidon.builder.Singular;
+import io.helidon.config.metadata.ConfiguredOption;
+import io.helidon.pico.DefaultBootstrap;
+import io.helidon.pico.types.DefaultTypeName;
+import io.helidon.pico.types.TypeName;
+
+/**
+ * Provides the basic formation for {@code module-info.java} creation and manipulation.
+ *
+ * @see java.lang.module.ModuleDescriptor
+ */
+@Builder
+public interface ModuleInfoDescriptor {
```

Comment:
Is this a parser of source of `module-info.java`?
There is class `java.lang.module.ModuleDescriptor` that supports reading the binary representation of module info - that can be used after compilation.
If you need to read module-info during compilation in annotation processor, I think there is a tool to obtain it correctly using annotation processor APIs.

This class seems redundant, and in general something that we should not be doing at all

Reply:
A few things to note on this one.

The java.lang.module.ModuleDescriptor is intended to work with the binary form of the module-info, not the source form. See https://docs.oracle.com/javase/9/docs/api/java/lang/module/ModuleDescriptor.html.  Same is true for its builder.

And fwiw, this is intended to be internally facing for tooling related to pico, at least that is the case initially. I can deprecate it perhaps to avoid common use perhaps?  I do agree it is unfortunate that I had to parse it but I found no other good alternatives considering there is no way to see or build the stringified version of the source.

Reply:
Follow-up task logged: https://github.com/helidon-io/helidon/issues/5697